src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs

```diff
@@ -0,0 +1,164 @@
+            settings.ConnectionString = connectionString;
+        }
+
+        configureSettings?.Invoke(settings);
+
+        if (settings.DbContextPooling)
+        {
+            builder.Services.AddDbContextPool<TContext>(ConfigureDbContext);
+        }
+        else
+        {
+            builder.Services.AddDbContext<TContext>(ConfigureDbContext);
+        }
+
+        if (settings.HealthChecks)
+        {
+            // calling MapHealthChecks is the responsibility of the app, not Component
+            builder.TryAddHealthCheck(
+                name: typeof(TContext).Name,
+                static hcBuilder => hcBuilder.AddDbContextCheck<TContext>());
+        }
+
+        if (settings.Tracing)
+        {
+            builder.Services.AddOpenTelemetry()
+                .WithTracing(tracerProviderBuilder =>
+                {
+                    // add tracing from the underlying MySqlConnector ADO.NET library
+                    tracerProviderBuilder.AddSource("MySqlConnector");
+                });
+        }
+
+        if (settings.Metrics)
+        {
+            builder.Services.AddOpenTelemetry()
+                .WithMetrics(meterProviderBuilder =>
+                {
+                    // Currently EF provides only Event Counters:
+                    // https://learn.microsoft.com/ef/core/logging-events-diagnostics/event-counters?tabs=windows#counters-and-their-meaning
+                    meterProviderBuilder.AddEventCountersInstrumentation(eventCountersInstrumentationOptions =>
+                    {
+                        // The magic strings come from:
+                        // https://github.com/dotnet/efcore/blob/a1cd4f45aa18314bc91d2b9ea1f71a3b7d5bf636/src/EFCore/Infrastructure/EntityFrameworkEventSource.cs#L45
+                        eventCountersInstrumentationOptions.AddEventSources("Microsoft.EntityFrameworkCore");
+                    });
+
+                    // add metrics from the underlying MySqlConnector ADO.NET library
+                    meterProviderBuilder.AddMeter("MySqlConnector");
+                });
+        }
+
+        void ConfigureDbContext(IServiceProvider serviceProvider, DbContextOptionsBuilder dbContextOptionsBuilder)
+        {
+            // use the legacy method of setting the ILoggerFactory because Pomelo EF Core doesn't use MySqlDataSource
+            if (serviceProvider.GetService<ILoggerFactory>() is { } loggerFactory)
+            {
+                MySqlConnectorLogManager.Provider = new MicrosoftExtensionsLoggingLoggerProvider(loggerFactory);
+            }
+
+            var connectionString = settings.ConnectionString ?? string.Empty;
+
+            ServerVersion serverVersion;
+            if (settings.ServerVersion is null)
+            {
+                if (string.IsNullOrEmpty(connectionString))
+                {
+                    ThrowForMissingConnectionString();
+                }
+                serverVersion = ServerVersion.AutoDetect(connectionString);
```

Comment:
Playing with this change locally, I'm a bit concerned about this now. I'm trying to run the integration tests (which aren't run in CI yet - we are working on that). If the Pomelo test is run by itself, it fails because just trying to get the DbContext out of DI is trying to establish a connection. But the MySql docker container isn't up yet (it takes a while to get up) and this line is throwing because it can't make a connection.

I wonder if we should retry this up to the MaxRetryCount. It isn't a great experience to fail getting the DbContext because it happens as part of ASP.NET calling into the user's code. So there's no chance for them to retry.

Thoughts? Maybe this shouldn't block the initial check-in, since you can work around the issue by setting ServerVersion explicitly.

Reply:
> I wonder if we should retry this up to the MaxRetryCount. 

And then still fail with a fatal error if it can't make a connection? (As opposed to making up a value like `"8.0.0"` or `"5.7.28"` or something that might not actually be safe if the user is running MariaDB, depending on how much this changes Pomelo's behavior.)

That definitely seems like an improvement over the status quo to me. (Unless of course 100% of the time it still fails but now just takes longer.)

In my original PR (before 2623f9322a6f81a22eb90ba0765f9f274f18258b), this was a mandatory option that had to be supplied by the user. That was less user-friendly on one hand, but maybe more predictable?

Very random thought (and may be going completely down the wrong path): is there some "magic" we could do by detecting if there's a `ContainerImageAnnotation`? E.g., set an appropriate default ServerVersion by faking a version string based on whether the user is using `mysql:latest` or `mariadb:11.0`? (That's not actually customisable in `AddMySqlContainer` right now though maybe it should be.) That only works for the local Docker scenario, but maybe that's OK because for production deployment we can assume that the "real" database is up and serving requests?

Reply:
Let's discuss the options in a new issue explicitly for this. Can you open it? or do you want me to?

Reply:
https://github.com/dotnet/aspire/issues/1866